web/src/elements/Base.ts

```diff
@@ -1,139 +1,152 @@
+    public override disconnectedCallback(): void {
         this.#themeAbortController?.abort();
+        super.disconnectedCallback();
     }
 
-    #styleRoot?: StyleSheetParent;
+    /**
+     * Returns the node into which the element should render.
+     *
+     * @see {LitElement.createRenderRoot} for more information.
+     */
+    protected override createRenderRoot(): HTMLElement | DocumentFragment {
+        const renderRoot = super.createRenderRoot();
+        this.styleRoot ??= renderRoot;
 
-    #dispatchTheme: UIThemeListener = (nextUITheme) => {
-        if (!this.#styleRoot) return;
+        return renderRoot;
+    }
 
-        if (nextUITheme === UiThemeEnum.Dark) {
-            this.#darkThemeStyleSheet ||= createStyleSheetUnsafe(ThemeDark);
-            appendStyleSheet(this.#styleRoot, this.#darkThemeStyleSheet);
-            this.activeTheme = UiThemeEnum.Dark;
-        } else if (this.#darkThemeStyleSheet) {
-            removeStyleSheet(this.#styleRoot, this.#darkThemeStyleSheet);
-            this.#darkThemeStyleSheet = null;
-            this.activeTheme = UiThemeEnum.Light;
-        }
-    };
+    //#endregion
 
-    protected createRenderRoot(): HTMLElement | DocumentFragment {
-        const renderRoot = super.createRenderRoot();
-        this.#styleRoot = resolveStyleSheetParent(renderRoot);
+    //#region Properties
 
-        if (this.#customCSSStyleSheet) {
-            console.debug(`authentik/element[${this.tagName.toLowerCase()}]: Adding custom CSS`);
+    /**
+     * The resolved theme of the current element.
+     *
+     * @remarks
+     *
+     * Unlike the browser's current color scheme, this is a value that can be
+     * resolved to a specific theme, i.e. dark or light.
+     */
+    @property({
+        attribute: "theme",
+        type: String,
+        reflect: true,
+    })
+    public activeTheme: ResolvedUITheme;
+
+    //#endregion
+
+    //#region Private Properties
+
+    /**
+     * The preferred color scheme used to look up the UI theme.
+     */
+    protected readonly preferredColorScheme: CSSColorSchemeValue;
+
+    /**
+     * A custom CSS style sheet to apply to the element.
+     */
+    readonly #customCSSStyleSheet: CSSStyleSheet | null;
+
+    /**
+     * A controller to abort theme updates, such as when the element is disconnected.
+     */
+    #themeAbortController: AbortController | null = null;
+    /**
+     * The style root to which the theme is applied.
+     */
+    #styleRoot?: StyleRoot;
 
-            appendStyleSheet(this.#styleRoot, this.#customCSSStyleSheet);
+    protected set styleRoot(nextStyleRoot: StyleRoot | undefined) {
+        this.#themeAbortController?.abort();
+
+        this.#styleRoot = nextStyleRoot;
+
+        if (!nextStyleRoot) return;
+
+        if (this.#customCSSStyleSheet) {
+            setAdoptedStyleSheets(nextStyleRoot, (currentStyleSheets) => {
+                return [...currentStyleSheets, this.#customCSSStyleSheet!];
+            });
```

Comment:
I suspect this touches a spec ambiguity and might come up in the future. Replacing a style root's `adoptedStyleSheets` will not take effect until the next repaint.

Appending to `adoptedStyleSheets` in the same frame would mean spreading out the previous entries, which would not yet be available in the array since they haven't yet mounted. We can't easily cache the previous entries since the style root may be a ShadyDOM.

That leaves us with a `requestAnimationFrame` to sequence the adoption + a visibility toggling to avoid a flash of styles between renders.

Reply:
@BeryJu @kensternberg-authentik

Reply:
ideally theming would just be a CSS file with a bunch of variables to set colours, but in PF4 there are still cases where colours are set statically/can't be overridden exactly enough